In [43]:
# Import necessary dependencies for data analysis and visualization
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [44]:
df: pd.DataFrame = pd.read_csv("..\\data\\titanic.csv")
df

PassengerId  Survived  Pclass  \
0            892         0       3   
1            893         1       3   
2            894         0       2   
3            895         0       3   
4            896         1       3   
..           ...       ...     ...   
413         1305         0       3   
414         1306         1       1   
415         1307         0       3   
416         1308         0       3   
417         1309         0       3   

                                             Name     Sex   Age  SibSp  Parch  \
0                                Kelly, Mr. James    male  34.5      0      0   
1                Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                       Myles, Mr. Thomas Francis    male  62.0      0      0   
3                                Wirz, Mr. Albert    male  27.0      0      0   
4    Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   
..                                            ...     ...   ...    ...    ...   
413                            Spector, Mr. Woolf    male   NaN      0      0   
414                  Oliva y Ocana, Dona. Fermina  female  39.0      0      0   
415                  Saether, Mr. Simon Sivertsen    male  38.5      0      0   
416                           Ware, Mr. Frederick    male   NaN      0      0   
417                      Peter, Master. Michael J    male   NaN      1      1   

                 Ticket      Fare Cabin Embarked  
0                330911    7.8292   NaN        Q  
1                363272    7.0000   NaN        S  
2                240276    9.6875   NaN        Q  
3                315154    8.6625   NaN        S  
4               3101298   12.2875   NaN        S  
..                  ...       ...   ...      ...  
413           A.5. 3236    8.0500   NaN        S  
414            PC 17758  108.9000  C105        C  
415  SOTON/O.Q. 3101262    7.2500   NaN        S  
416              359309    8.0500   NaN        S  
417                2668   22.3583   NaN        C  

[418 rows x 12 columns]

## Data Preparing Stage

- Trying to get an overview about the dataset to indentify key concepts, problems, relations...

In [45]:
# Display the first 5 rows of the dataset
df.head()

PassengerId  Survived  Pclass  \
0          892         0       3   
1          893         1       3   
2          894         0       2   
3          895         0       3   
4          896         1       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                              Kelly, Mr. James    male  34.5      0      0   
1              Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                     Myles, Mr. Thomas Francis    male  62.0      0      0   
3                              Wirz, Mr. Albert    male  27.0      0      0   
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   

    Ticket     Fare Cabin Embarked  
0   330911   7.8292   NaN        Q  
1   363272   7.0000   NaN        S  
2   240276   9.6875   NaN        Q  
3   315154   8.6625   NaN        S  
4  3101298  12.2875   NaN        S

In [46]:
# Display basic information about the dataset (columns, non-null counts, data types)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Survived     418 non-null    int64  
 2   Pclass       418 non-null    int64  
 3   Name         418 non-null    object 
 4   Sex          418 non-null    object 
 5   Age          332 non-null    float64
 6   SibSp        418 non-null    int64  
 7   Parch        418 non-null    int64  
 8   Ticket       418 non-null    object 
 9   Fare         417 non-null    float64
 10  Cabin        91 non-null     object 
 11  Embarked     418 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 39.3+ KB


In [47]:
# Display summary statistics for numerical columns
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   418.000000  418.000000  418.000000  332.000000  418.000000   
mean   1100.500000    0.363636    2.265550   30.272590    0.447368   
std     120.810458    0.481622    0.841838   14.181209    0.896760   
min     892.000000    0.000000    1.000000    0.170000    0.000000   
25%     996.250000    0.000000    1.000000   21.000000    0.000000   
50%    1100.500000    0.000000    3.000000   27.000000    0.000000   
75%    1204.750000    1.000000    3.000000   39.000000    1.000000   
max    1309.000000    1.000000    3.000000   76.000000    8.000000   

            Parch        Fare  
count  418.000000  417.000000  
mean     0.392344   35.627188  
std      0.981429   55.907576  
min      0.000000    0.000000  
25%      0.000000    7.895800  
50%      0.000000   14.454200  
75%      0.000000   31.500000  
max      9.000000  512.329200

In [48]:
# Check for missing values
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

In [49]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

### Initial Observations
- The dataset contains **418 rows** and **12 columns**.
- **Columns include**: `PassengerId`, `Survived`, `Pclass`, `Name`, `Sex`, `Age`, `SibSp`, `Parch`, `Ticket`, `Fare`, `Cabin`, `Embarked`.
- **Missing values** are present in the following columns:
    - `Age`: 86 missing values (20.6% of the data).
    - `Fare`: 1 missing value (0.2% of the data).
    - `Cabin`: 327 missing values (78.2% of the data).
- **Summary Statistics** reveal:
    - **36.4% of passengers survived**.
    - The majority of passengers were in **third class** (Pclass = 3).
    - The age of passengers ranges from **0 to 76 years**, with an **average age of 30 years**.
    - Most passengers traveled **without siblings or spouses** (75% had 0 SibSp).
    - Most passengers traveled **without parents or children** (75% had 0 Parch).
    - The fare ranges from **0 to 512.33 pounds**, with a median of **14.45 pounds**.
- **Data types**:
    - **Categorical Columns**: `Survived`, `Pclass`, `Sex`, `Cabin`, `Embarked` are categorical and should be treated as such.
    - **Numerical Columns**: `Age`, `Fare`, `SibSp`, `Parch` are numerical and can be used for statistical analysis.
    - **Text Columns**: `Name` and `Ticket` are text-based and may require special handling (e.g., extracting titles from names).

## Data Cleaning Stage
Since the dataset contains missing values I need to handle them for each columns using the following aproaches:
- **`Age`**: Imputing missing values using the mean.
- **`Fare`**: Single missing value can be imputed using the median fare.
- **`Cabin`**: The high percentage of missing values (78.2%) suggests that this column may not be useful for analysis unless additional information is available.

In [50]:
age_mean: float = df['Age'].mean().__round__(2)
age_mean

np.float64(30.27)

In [51]:
df['Age'].fillna(age_mean, inplace=True)
df

C:\Users\dell\AppData\Local\Temp\ipykernel_15324\3644991198.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Age'].fillna(age_mean, inplace=True)


PassengerId  Survived  Pclass  \
0            892         0       3   
1            893         1       3   
2            894         0       2   
3            895         0       3   
4            896         1       3   
..           ...       ...     ...   
413         1305         0       3   
414         1306         1       1   
415         1307         0       3   
416         1308         0       3   
417         1309         0       3   

                                             Name     Sex    Age  SibSp  \
0                                Kelly, Mr. James    male  34.50      0   
1                Wilkes, Mrs. James (Ellen Needs)  female  47.00      1   
2                       Myles, Mr. Thomas Francis    male  62.00      0   
3                                Wirz, Mr. Albert    male  27.00      0   
4    Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.00      1   
..                                            ...     ...    ...    ...   
413                            Spector, Mr. Woolf    male  30.27      0   
414                  Oliva y Ocana, Dona. Fermina  female  39.00      0   
415                  Saether, Mr. Simon Sivertsen    male  38.50      0   
416                           Ware, Mr. Frederick    male  30.27      0   
417                      Peter, Master. Michael J    male  30.27      1   

     Parch              Ticket      Fare Cabin Embarked  
0        0              330911    7.8292   NaN        Q  
1        0              363272    7.0000   NaN        S  
2        0              240276    9.6875   NaN        Q  
3        0              315154    8.6625   NaN        S  
4        1             3101298   12.2875   NaN        S  
..     ...                 ...       ...   ...      ...  
413      0           A.5. 3236    8.0500   NaN        S  
414      0            PC 17758  108.9000  C105        C  
415      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416      0              359309    8.0500   NaN        S  
417      1                2668   22.3583   NaN        C  

[418 rows x 12 columns]

In [52]:
median_fare: float = df['Fare'].median()
median_fare

np.float64(14.4542)

In [53]:
df['Fare'].fillna(median_fare, inplace=True)
df

C:\Users\dell\AppData\Local\Temp\ipykernel_15324\2725541655.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Fare'].fillna(median_fare, inplace=True)


PassengerId  Survived  Pclass  \
0            892         0       3   
1            893         1       3   
2            894         0       2   
3            895         0       3   
4            896         1       3   
..           ...       ...     ...   
413         1305         0       3   
414         1306         1       1   
415         1307         0       3   
416         1308         0       3   
417         1309         0       3   

                                             Name     Sex    Age  SibSp  \
0                                Kelly, Mr. James    male  34.50      0   
1                Wilkes, Mrs. James (Ellen Needs)  female  47.00      1   
2                       Myles, Mr. Thomas Francis    male  62.00      0   
3                                Wirz, Mr. Albert    male  27.00      0   
4    Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.00      1   
..                                            ...     ...    ...    ...   
413                            Spector, Mr. Woolf    male  30.27      0   
414                  Oliva y Ocana, Dona. Fermina  female  39.00      0   
415                  Saether, Mr. Simon Sivertsen    male  38.50      0   
416                           Ware, Mr. Frederick    male  30.27      0   
417                      Peter, Master. Michael J    male  30.27      1   

     Parch              Ticket      Fare Cabin Embarked  
0        0              330911    7.8292   NaN        Q  
1        0              363272    7.0000   NaN        S  
2        0              240276    9.6875   NaN        Q  
3        0              315154    8.6625   NaN        S  
4        1             3101298   12.2875   NaN        S  
..     ...                 ...       ...   ...      ...  
413      0           A.5. 3236    8.0500   NaN        S  
414      0            PC 17758  108.9000  C105        C  
415      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416      0              359309    8.0500   NaN        S  
417      1                2668   22.3583   NaN        C  

[418 rows x 12 columns]

In [56]:
df.drop(columns='Cabin', inplace=True)
df

PassengerId  Survived  Pclass  \
0            892         0       3   
1            893         1       3   
2            894         0       2   
3            895         0       3   
4            896         1       3   
..           ...       ...     ...   
413         1305         0       3   
414         1306         1       1   
415         1307         0       3   
416         1308         0       3   
417         1309         0       3   

                                             Name     Sex    Age  SibSp  \
0                                Kelly, Mr. James    male  34.50      0   
1                Wilkes, Mrs. James (Ellen Needs)  female  47.00      1   
2                       Myles, Mr. Thomas Francis    male  62.00      0   
3                                Wirz, Mr. Albert    male  27.00      0   
4    Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.00      1   
..                                            ...     ...    ...    ...   
413                            Spector, Mr. Woolf    male  30.27      0   
414                  Oliva y Ocana, Dona. Fermina  female  39.00      0   
415                  Saether, Mr. Simon Sivertsen    male  38.50      0   
416                           Ware, Mr. Frederick    male  30.27      0   
417                      Peter, Master. Michael J    male  30.27      1   

     Parch              Ticket      Fare Embarked  
0        0              330911    7.8292        Q  
1        0              363272    7.0000        S  
2        0              240276    9.6875        Q  
3        0              315154    8.6625        S  
4        1             3101298   12.2875        S  
..     ...                 ...       ...      ...  
413      0           A.5. 3236    8.0500        S  
414      0            PC 17758  108.9000        C  
415      0  SOTON/O.Q. 3101262    7.2500        S  
416      0              359309    8.0500        S  
417      1                2668   22.3583        C  

[418 rows x 11 columns]